Homework2

marl0vv opened the pull request:
- Добавил зависимости для String Boost. Зависимости для lombok и Gson были с прошлого задания.
- lombok был добавлен в прошлой работе. Сгенерировал геттеры, сеттеры и конструкторы для класса `Entry` Не совсем понял про проверки для обязательных аргументов методов
- Добавил в класс Storage метку @ Component. Если я правильно понял, то это и означает сделать её бином.
- Путь до файла содержится в application.properties, считывается через аннотацию @ Value в `String` , который затем передаётся в функцию `initializeStorage ` 
- Я работал с json ещё в прошлом задании, не знал про Jackson и ObjectMapper, поэтому использовал гугловский Gson. Если это важно, то могу переписать.
- Добавил к классу Service implements CommandLineRunner и аннотацию @ Override на функцию `run()`
- Все тесты проходят. Не совсем уверен, как нужно было получать файл с результатами, я скопировал вывод из консоли в .txt файл
[mvn clean package output.txt](https://github.com/marl0vv/whitesoft-2023-backend-storage/files/13205213/mvn.clean.package.output.txt)

marl0vv committed:
Added dependencies

Added Spring Boot dependencies
There are also lombok dependencies in the file, I added them in previous homework.

marl0vv committed:
Added Spring Boot

Storage class is now bean
Path to inputData.json is now read from application.properties
Program menu now starts with CommandLineRunner

marl0vv committed:
Deleted JUnit dependencies

I'm not sure but It seems that Spring Boot has its own JUnit dependencies or something and they got in conflict with dependencies that I had earlier. Anyway I deleted them and it's working now.

marl0vv committed:
fix test

It seems I broke one of my tests in previous homework when I change Entry from Record into Class. Asserting equals didn't work because it just compared references by itself. Overrode equals method to compare them in a right way.

marl0vv committed:
formatting

Konast376 commented on pom.xml:
```diff
@@ -46,10 +38,42 @@
             <scope>provided</scope>
         </dependency>
 
+        <dependency>
```
вместо двух зависимостей ниже можно использовать 

```
<dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter</artifactId>
            <version>3.1.5</version>
</dependency>
```

marl0vv replied:
`spring-boot-starter` позволил заменить` spring-boot-starter-web`, но не позволил заменить `spring-boot-starter-test`. Проект просто не увидел JUnit и не запустил тесты. Возможно, я не так понял. Пока не стал ничего менять.

Konast376 commented on src/main/java/com/marl0vv/study/Entry.java:
```diff
@@ -1,15 +1,12 @@
 @NoArgsConstructor
 @AllArgsConstructor
-public class Entry{
+public class Entry {
```
Стоит дать классу название, соответствующее логике

marl0vv replied:
Можно подсказку по наименованию? Потому что этот класс являет собой одну из записей, которые хранятся в хранилище Storage, поэтому и назвал Entry. Возможно, тогда StorageItem подойдёт?

Konast376 replied:
Подойдет

Konast376 commented on src/main/java/com/marl0vv/study/Input.java:
```diff
@@ -35,12 +35,14 @@ private static void addEntriesToStorage(List<Entry> entries, Map<Integer, Entry>
+
     public static Map<Integer, Entry> readFromFile(String filePath) {
         Map<Integer, Entry> storage = new HashMap<>();
         Gson gson = new Gson();
```
По заданию для работы с json нужно использовать ObjectMapper

marl0vv replied:
Готово. Заменил Gson на ObjectMapper 👍

Konast376 commented on pom.xml:
```diff
@@ -14,21 +14,13 @@
-            <version>5.10.0</version>
-            <scope>test</scope>
-        </dependency>
+    <parent>
```
Можно убрать

Konast376 commented on src/main/java/com/marl0vv/study/Main.java:
```diff
@@ -1,9 +1,28 @@
+
+    @Bean
+    @Order(Ordered.HIGHEST_PRECEDENCE)
+    public CommandLineRunner setupData(Storage storage) {
```
добавь отдельный класс для этого. можешь назвать его прям AppRunner. унаследуй его от implements CommandLineRunner. повесь на него @Component. и вот в переопределяющем методе run сделай всю логику

из класса Main убери вообще всё. должно остаться только:

```
@SpringBootApplication
public class Main {
    public static void main(String[] args) {
        SpringApplication.run(Main.class, args);
    }
}
```

типа как [здесь](https://www.baeldung.com/spring-boot-console-app), но вынесено в другой класс, а не налеплено всё в Main.main

marl0vv replied:
Готово. Убрал всю логику из Main в AppRunner 👍

Konast376 commented on src/main/java/com/marl0vv/study/Entry.java:
```diff
@@ -18,4 +15,20 @@ public class Entry{
     }
+
+    @Override
+    public boolean equals(Object object) {
```
При переопределении equals нужно также переопределять hashCode. В данном случае можно заменить написание этих методов, а также аннотации Getter и Setter на аннотацию Data

marl0vv replied:
Готово 👍

Konast376 commented on src/main/java/com/marl0vv/study/Storage.java:
```diff
@@ -1,22 +1,30 @@
     private Map<Integer, Entry> storage;
-    public Storage(String filePath){
+
+    public Storage() {
```
Можно использовать аннотацию NoArgsConsrtuctor

marl0vv replied:
Сделано 👍

Konast376 commented on src/main/java/com/marl0vv/study/Storage.java:
```diff
@@ -1,22 +1,30 @@
 import java.util.List;
 import java.util.Map;
+
+@Component
```
Эту штуку не стоит делать компонентом.

marl0vv replied:
Тогда я, видимо, что-то не понимаю. В задании было сказано сделать бином класс, отвечающий за работу с хранилищем, поэтому я туда аннотацию Component и повесил. Вдобавок, если я убираю аннотацию Component тогда в Service и AppRunner я получаю ошибку "No beans of 'Storage' type found".

Тогда образуются следующие вопросы:
1. Почему эту штуку не нужно делать компонентой?
2. Как избавиться от ошибки?
3. Где можно про это почитать?

Konast376 replied:
Так, здесь, кажется, я ошибся. Все ок

Konast376 commented on src/test/java/com/marl0vv/study/ServiceTest.java:
```diff
@@ -22,26 +28,26 @@ void testFindById(){
 
         assertNotNull(matchingEntries);
-        for(Entry entry : matchingEntries){
+        for (Entry entry : matchingEntries) {
```
В тестах лучше не использовать циклы. Если проверяешь коллекцию, то сначала пишешь assert на количество элементов в ней, а потом проверяешь содержание каждого элемента

marl0vv replied:
Это чертовски хороший совет! Спасибо. Оказалось, что у меня была ошибка в коде. Оказывается, что я приводил к нижнему регистру пользовательскую строку прямо во время работы функции `findEntry` в классе `Service`,  но так как во время тестов я передавал строку не через `Service`, а напрямую в функцию `findByName`, у меня не находилось совпадений.

Тест не выдавал ошибку, так как цикл просто ни разу не запускался и не доходил до assert.

Всё поправил.

Konast376 commented on src/test/java/com/marl0vv/study/ServiceTest.java:
```diff
@@ -1,19 +1,25 @@
 import static org.junit.jupiter.api.Assertions.*;
 
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
 public class ServiceTest {
```
Тесты побелить комментариями на блоки Arrange, Act, Assert

marl0vv replied:
Готово 👍

Konast376 commented on src/test/java/com/marl0vv/study/ServiceTest.java:
```diff
@@ -1,19 +1,25 @@
 
 import static org.junit.jupiter.api.Assertions.*;
 
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
```
Вместо этой аннотации и BeforeAll можно использовать BeforeEach. Но так тоже норм :)

marl0vv replied:
👍

Konast376 commented on src/main/java/com/marl0vv/study/Service.java:
```diff
@@ -1,14 +1,19 @@
 
-public class Service {
+@Component
+public class Service implements CommandLineRunner {
```
Нужно сделать Service отдельно от CommandLineRunner, они сами по себе

marl0vv replied:
Готово 👍

marl0vv committed:
Change Gson to Jackson. Using ObjectMapper

Deleted Gson dependency in pom.xml and changed it with Jackson.
Using ObjectMapper instead of Gson in Input.java

marl0vv committed:
Created AppRunner class

Moved all logic from Main to AppRunner

marl0vv committed:
Changed equals() method with lombok version

marl0vv committed:
Lombok NoArgsConstructor used

marl0vv committed:
Added comments with Arrange, Act, Assert in tests

marl0vv committed:
Deleted loop from the testFindByName()
